doc/Architecture/Subtyping.org

```diff
@@ -0,0 +1,146 @@
+* Subtyping
+** An argument about the extent of subtyping
+
+Subtyping is an important part of any module or Object system. The
+rules these systems usually care about revolve around the subtype of
+=Records= and =Records= in relation to functions.
+
+*** About Subtyping Rules on Records
+An example of a record subtyping is as follows, we will use =a :> b= to
+mean that =a= is a subtype of =b=
+
+ 1. ∀x, a \dots v : Type. {x : Type, a : Type \dots v : Type} :> {a \dots v : Type}
+ 2. ∀x, a \dots v : Type. ∀ f : {a : Type \dots v : Type} → Type \\
+    f :> f : {x : Type, a : Type \dots v : Type} → Type
+
+What 1 tells us is that we can /forget/ a field, and thus a record
+with more fields is a sub type of the same record with less fields
+
+What 2 tells us is that a function which accepts a record with fields
+=a= through =v= can also accept a record with those fields /and/ extra
+fields. The intuition is that in functions the sub typing relation is
+reversed, but we still need to forget information to get our
+relationship.
+
+
+*** Implications of Rules and Arguments about Behavior
+
+These implications can be quite worrying, consider the longer form
+example below. We will introduce all the forms first then make an
+argument about what various object systems have done and what we are
+proposing to do.
+
+#+begin_src ocaml
+  f0 : λX:>{x : int} -> x1 -> x2 -> x1
+  f0 rec1 rec2 = rec1 { x = rec1.x + rec2.x}
+
+
+  f1 : λX:>{x : int} -> X -> X -> X
+  f1 rec1 rec2 = rec1 {x = rec1.x + rec2.x}
+
+
+  f2 : λX:>{x : int} -> X -> X -> X
+  f2 rec1 rec2 = {x = rec1.x + rec2.x}
+
+
+  -- note that the t is opaque
+  sig set : int -> mod {x : t,  type t, add : t -> t -> t}
+  mod set x =
+    type t = int
+    add = (+)
+    x
+  end
+
+  -- We include the set
+  sig set : int -> mod {x : t,  type t, add : t -> t -> t}
+  mod bag x = include (set x)
```

Comment:
is this valid syntax?

Reply:
signature syntax isn't fleshed out yet, but the rest should be?

Reply:
I removed the mod word